File: agent/shopping_normalizer.py
"""
Shopping List Normalizer Agent

Purpose: Intelligently normalize and aggregate ingredients from multiple recipes
into a clean, consolidated shopping list.

Input: Array of ingredient objects from multiple recipes
Output: Normalized array with aggregated quantities
"""

from google.adk.agents import Agent
from google.adk.models.google_llm import Gemini
from google.adk.runners import InMemoryRunner
from google.genai import types
import json
import sys


# Agent configuration
normalizer_agent = Agent(
    name="ShoppingListNormalizer",
    model=Gemini(model_name="gemini-2.0-flash-exp"),
    system_instruction="""You are an expert shopping list normalizer. Your job is to take raw ingredient lists from multiple recipes and produce a clean, consolidated shopping list.

## Instructions:

1. **Normalize ingredient names**: Convert to lowercase, standardize spelling, group similar items
   - "Chickpeas" → "chickpeas"
   - "chick peas" → "chickpeas"
   - "garbanzo beans" → "chickpeas"
   
2. **Aggregate quantities intelligently**:
   - Whole units: "2 cans + 1 can" → "3 cans"
   - Volumes: "1 cup + 2 cups" → "3 cups"
   - Weights: "8 oz + 4 oz" → "12 oz"
   - Mixed units: Keep separate entries if conversion is ambiguous
   
3. **Extract notes from context**:
   - "canned chickpeas" → name: "chickpeas", notes: "canned"
   - "fresh basil" → name: "basil", notes: "fresh"
   - "diced tomatoes" → name: "tomatoes", notes: "diced"
   
4. **Handle common variations**:
   - Singular/plural: "tomato" and "tomatoes" → "tomatoes"
   - Different forms: "olive oil", "oil (olive)" → "olive oil"
   - Brand specifications: Ignore or move to notes
   
5. **Smart grouping**:
   - Group similar preparations: "diced tomatoes" + "crushed tomatoes" → "tomatoes (diced, crushed)"
   - Keep distinct items separate: "fresh tomatoes" ≠ "canned tomatoes"

## Output Format:

Return ONLY a JSON array with this exact structure:
```json
[
  {
    "name": "chickpeas",
    "qty": "3 cans",
    "notes": "canned"
  },
  {
    "name": "olive oil",
    "qty": "1/2 cup",
    "notes": ""
  }
]
```

## Rules:
- Return ONLY valid JSON, no explanations
- Each item must have "name", "qty", and "notes" fields
- Use lowercase for names
- Combine quantities when units match
- Notes should be concise (e.g., "fresh", "canned", "diced")
- If quantity is unclear, use "as needed" or "to taste"
- Sort alphabetically by name for easy shopping""",
    generation_config=types.GenerateContentConfig(
        temperature=0.3,
        response_mime_type="application/json"
    )
)


def normalize_shopping_list(ingredients_data):
    """
    Normalize and aggregate a shopping list from multiple recipes.
    
    Args:
        ingredients_data: List of ingredient objects with 'name', 'qty', etc.
        
    Returns:
        Normalized list of ingredients as JSON
    """
    # Convert input to a readable format for the LLM
    ingredients_text = json.dumps(ingredients_data, indent=2)
    
    prompt = f"""Normalize and aggregate this shopping list from multiple recipes:

{ingredients_text}

Return a consolidated shopping list with aggregated quantities, normalized names, and extracted notes."""
    
    runner = InMemoryRunner()
    result = runner.run(agent=normalizer_agent, prompt=prompt)
    
    try:
        # Parse the JSON response
        normalized_list = json.loads(result)
        return normalized_list
    except json.JSONDecodeError as e:
        print(f"Error parsing agent response: {e}", file=sys.stderr)
        print(f"Raw response: {result}", file=sys.stderr)
        return []


def main():
    """
    Main entry point - can accept input from stdin or use sample data.
    """
    import sys
    
    # Check if input is provided via stdin
    if not sys.stdin.isatty():
        try:
            # Read from stdin (when called from Node.js)
            input_data = sys.stdin.read()
            ingredients_data = json.loads(input_data)
            
            # Normalize and output only the result
            normalized = normalize_shopping_list(ingredients_data)
            print(json.dumps(normalized, indent=2))
            return normalized
            
        except Exception as e:
            print(f"Error processing stdin: {e}", file=sys.stderr)
            sys.exit(1)
    
    # Test mode with sample data
    sample_ingredients = [
        {"name": "Chickpeas", "qty": "1 can"},
        {"name": "chickpeas", "qty": "2 cans"},
        {"name": "Olive Oil", "qty": "2 tbsp"},
        {"name": "olive oil", "qty": "1 tbsp"},
        {"name": "Fresh Basil", "qty": "1 cup"},
        {"name": "basil", "qty": "1/2 cup", "item": "fresh"},
        {"name": "Diced Tomatoes", "qty": "1 can (14 oz)"},
        {"name": "tomatoes", "qty": "2 medium", "item": "fresh tomatoes"},
        {"name": "Garlic", "qty": "3 cloves"},
        {"name": "garlic cloves", "qty": "2"},
        {"name": "Quinoa", "qty": "1 cup"},
        {"name": "quinoa", "qty": "1/2 cup"},
        {"name": "Lemon", "qty": "1"},
        {"name": "lemon juice", "qty": "2 tbsp"},
        {"name": "Red Bell Pepper", "qty": "1 large"},
        {"name": "bell peppers", "qty": "2 small", "item": "red"},
        {"name": "Salt", "qty": "to taste"},
        {"name": "salt", "qty": "1 tsp"},
        {"name": "Black Pepper", "qty": "to taste"},
        {"name": "Cumin", "qty": "1 tsp"},
        {"name": "ground cumin", "qty": "1/2 tsp"}
    ]
    
    print("=== SHOPPING LIST NORMALIZER TEST ===\n", file=sys.stderr)
    print("Input ingredients:", file=sys.stderr)
    print(json.dumps(sample_ingredients, indent=2), file=sys.stderr)
    print("\n" + "="*60 + "\n", file=sys.stderr)
    
    # Normalize the list
    normalized = normalize_shopping_list(sample_ingredients)
    
    print("Normalized shopping list:", file=sys.stderr)
    print(json.dumps(normalized, indent=2))
    
    return normalized


if __name__ == "__main__":
    main()

File: agent/shoppingNormalizer.js
/**
 * Shopping List Normalizer - Node.js Wrapper
 * 
 * Wraps the Python ADK shopping normalizer agent for use in Node.js services.
 * Can be used by the shoppingListTool to produce cleaner, aggregated lists.
 */

const { spawn } = require('child_process');
const path = require('path');

/**
 * Normalize a shopping list using the AI agent
 * 
 * @param {Array} ingredients - Array of ingredient objects with name, qty, etc.
 * @returns {Promise<Array>} - Normalized and aggregated shopping list
 */
async function normalizeShoppingList(ingredients) {
  return new Promise((resolve, reject) => {
    const pythonScript = path.join(__dirname, 'shopping_normalizer.py');
    const python = spawn('python3', [pythonScript]);
    
    let stdout = '';
    let stderr = '';
    
    // Send ingredients as JSON to Python script via stdin
    python.stdin.write(JSON.stringify(ingredients));
    python.stdin.end();
    
    python.stdout.on('data', (data) => {
      stdout += data.toString();
    });
    
    python.stderr.on('data', (data) => {
      stderr += data.toString();
    });
    
    python.on('close', (code) => {
      if (code !== 0) {
        console.error('Python agent error:', stderr);
        // Fallback to basic normalization
        resolve(basicNormalization(ingredients));
        return;
      }
      
      try {
        // Extract JSON from stdout (may have other output)
        const jsonMatch = stdout.match(/\[[\s\S]*\]/);
        if (jsonMatch) {
          const normalized = JSON.parse(jsonMatch[0]);
          resolve(normalized);
        } else {
          console.warn('No JSON found in agent output, using basic normalization');
          resolve(basicNormalization(ingredients));
        }
      } catch (err) {
        console.error('Error parsing agent response:', err.message);
        resolve(basicNormalization(ingredients));
      }
    });
    
    // Timeout after 30 seconds
    setTimeout(() => {
      python.kill();
      console.warn('Agent timeout, using basic normalization');
      resolve(basicNormalization(ingredients));
    }, 30000);
  });
}

/**
 * Basic normalization fallback (if AI agent fails)
 * Simple grouping by lowercase name
 */
function basicNormalization(ingredients) {
  const grouped = new Map();
  
  ingredients.forEach(ing => {
    const name = (ing.name || ing.item || '').toLowerCase().trim();
    if (!name) return;
    
    const qty = ing.qty || ing.quantity || '1 unit';
    
    if (grouped.has(name)) {
      const existing = grouped.get(name);
      existing.quantities.push(qty);
    } else {
      grouped.set(name, {
        name: name,
        quantities: [qty],
        notes: ''
      });
    }
  });
  
  // Convert to output format
  return Array.from(grouped.values())
    .map(item => ({
      name: item.name,
      qty: item.quantities.join(', '),
      notes: item.notes
    }))
    .sort((a, b) => a.name.localeCompare(b.name));
}

/**
 * Enhanced normalization that can be used directly without AI
 * (for when Python/AI is not available)
 */
function enhancedNormalization(ingredients) {
  const grouped = new Map();
  
  ingredients.forEach(ing => {
    let name = (ing.name || ing.item || '').toLowerCase().trim();
    const qty = ing.qty || ing.quantity || '1 unit';
    let notes = '';
    
    // Extract notes from name
    const notePatterns = [
      { pattern: /\b(fresh|dried|frozen|canned|crushed|diced|chopped|minced|sliced)\b/gi, type: 'prep' },
      { pattern: /\b(ground|whole|raw)\b/gi, type: 'form' }
    ];
    
    notePatterns.forEach(({ pattern }) => {
      const matches = name.match(pattern);
      if (matches) {
        notes = matches.join(', ').toLowerCase();
        matches.forEach(match => {
          name = name.replace(new RegExp(`\\b${match}\\b`, 'gi'), '').trim();
        });
      }
    });
    
    // Clean up extra spaces
    name = name.replace(/\s+/g, ' ').trim();
    
    // Normalize common variations
    const normalizations = {
      'chickpeas': ['garbanzo beans', 'garbanzos', 'chick peas'],
      'bell pepper': ['bell peppers', 'sweet pepper', 'sweet peppers'],
      'olive oil': ['oil (olive)', 'extra virgin olive oil'],
      'tomato': ['tomatoes'],
      'onion': ['onions'],
      'garlic': ['garlic clove', 'garlic cloves', 'clove of garlic', 'cloves garlic']
    };
    
    for (const [standard, variations] of Object.entries(normalizations)) {
      if (variations.some(v => name.includes(v))) {
        name = standard;
        break;
      }
    }
    
    const key = `${name}-${notes}`;
    
    if (grouped.has(key)) {
      const existing = grouped.get(key);
      existing.quantities.push(qty);
    } else {
      grouped.set(key, {
        name,
        quantities: [qty],
        notes
      });
    }
  });
  
  // Convert to output format
  return Array.from(grouped.values())
    .map(item => ({
      name: item.name,
      qty: item.quantities.length > 1 
        ? item.quantities.join(' + ') 
        : item.quantities[0],
      notes: item.notes
    }))
    .sort((a, b) => a.name.localeCompare(b.name));
}

module.exports = {
  normalizeShoppingList,
  basicNormalization,
  enhancedNormalization
};

